core/src/main/java/org/apache/iceberg/io/ResolvingFileIO.java

```diff
@@ -0,0 +1,153 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.iceberg.io;
+
+import java.util.List;
+import java.util.Map;
+import java.util.function.Function;
+import org.apache.hadoop.conf.Configuration;
+import org.apache.iceberg.CatalogUtil;
+import org.apache.iceberg.hadoop.HadoopConfigurable;
+import org.apache.iceberg.hadoop.SerializableConfiguration;
+import org.apache.iceberg.relocated.com.google.common.collect.ImmutableMap;
+import org.apache.iceberg.relocated.com.google.common.collect.Lists;
+import org.apache.iceberg.relocated.com.google.common.collect.Maps;
+import org.apache.iceberg.util.SerializableSupplier;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * FileIO implementation that uses location scheme to choose the correct FileIO implementation.
+ */
+public class ResolvingFileIO implements FileIO, HadoopConfigurable {
+  private static final Logger LOG = LoggerFactory.getLogger(ResolvingFileIO.class);
+  private static final String DEFAULT_SCHEME = "fs";
+  private static final String FALLBACK_IMPL = "org.apache.iceberg.hadoop.HadoopFileIO";
+  private static final Map<String, String> SCHEME_TO_FILE_IO = ImmutableMap.of(
```

Comment:
Having this mapping hardcoded feels a little inflexible for future FileIO implementations (open or custom implementations) because we will have to do a release to add new ones.

Reply:
I agree, but not having this hard coded opens up a lot more concerns. If we allow customizing this, how is that configured? It seems like it would probably be done through catalog properties. What are those properties? Do we support aliasing s3/s3a/s3n?

I'm trying to avoid creating a lot of custom config here, but it may be better to simply design for this from the start. But that's a longer path with more questions.